README.rst
=======
svmplus
=======


.. image:: https://img.shields.io/pypi/v/svmplus.svg
        :target: https://pypi.python.org/pypi/svmplus

.. image:: https://img.shields.io/travis/niharikag/svmplus.svg
        :target: https://travis-ci.org/niharikag/svmplus

.. image:: https://readthedocs.org/projects/svmplus/badge/?version=latest
        :target: https://svmplus.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status




Python implementation of Support Vector Machines Plus (SVM+).
Implementation of Support Vector Machines Plus (SVM+) for classification problems.
See [1] for theoretical details and see [2] for implementation details in 'MATLAB'.

This can be considered as an extension to sklearn.svm.


* Free software: MIT license
* Documentation: https://svmplus.readthedocs.io.


Features
--------

* TODO

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

1.0.0 (2018-05-05)
------------------

* First release on PyPI.


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for svmplus."""


tests/test_svmplus.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `svmplus` package."""


import unittest

from svmplus import svmplus


class TestSvmplus(unittest.TestCase):
    """Tests for `svmplus` package."""

    def setUp(self):
        """Set up test fixtures, if any."""

    def tearDown(self):
        """Tear down test fixtures, if any."""

    def test_000_something(self):
        """Test something."""


examples/test_svmplus.py
import numpy as np
import svmplus
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
import scipy.ndimage


def prepareDigitData():
    digits = load_digits(n_class=2)
    X = digits.data
    y = digits.target
    y[y == 0] = -1

    resizedImage = np.zeros((len(X), 16))
    for i in range(len(X)):
        originalImage = X[i].reshape(8, 8)
        originalImage = scipy.ndimage.zoom(originalImage, 4 / (8), order=0)
        resizedImage[i] = originalImage.reshape(16, )

    XStar = X[:]
    X = resizedImage

    X_train, X_test, y_train, y_test, indices_train, indices_valid = \
        train_test_split(X, y, range(len(X)), test_size=0.3, stratify = y)

    XStar = XStar[indices_train]
    return X_train, X_test, y_train, y_test, XStar


def testLinearSVMPlus():
    X_train, X_test, y_train, y_test, XStar = prepareDigitData()

    svmp = svmplus.SVMPlus(svm_type="QP", C=1000, gamma = .00001, kernel_x="linear",
                   kernel_xstar="linear", tol = 1e-10)
    svmp.fit(X_train, XStar, np.array(y_train).reshape(len(X_train),1))
    y_predict = svmp.predict(X_test)
    correct = np.sum(y_predict == y_test)
    print("Prediction accuracy of SVM")
    print("%d out of %d predictions correct" % (correct, len(y_predict)))


def testPolynomialSVMPlus():
    X_train, X_test, y_train, y_test, XStar = prepareDigitData()

    svmp = svmplus.SVMPlus(svm_type="QP", C=10, gamma = .01, kernel_x="poly",
                   kernel_xstar="poly", tol = 1e-10)
    svmp.fit(X_train, XStar, np.array(y_train).reshape(len(X_train),1))
    y_predict = svmp.predict(X_test)
    correct = np.sum(y_predict == y_test)
    print("Prediction accuracy of SVM")
    print("%d out of %d predictions correct" % (correct, len(y_predict)))


def testGridSerachCV():
    param_grid = {'C': [10, 100],
                  'gamma': [0.0001, 0.001],
                  'gamma_x': [0.0001, 0.001],
                  'gamma_xstar': [0.0001, 0.001]}

    X_train, X_test, y_train, y_test, XStar = prepareDigitData()

    c, gamma, gamma_x, gamma_x_star = gridSearchCV.gridSearchSVMPlus(X_train, y_train, XStar,
                                   param_grid, n_splits=5, logfile=None)
    print(c, gamma, gamma_x, gamma_x_star)



def testRbfSVMPlus():
    X_train, X_test, y_train, y_test, XStar = prepareDigitData()

    # train and predict using SVM plus
    svmp = svmplus.SVMPlus(svm_type="QP", C=1000, gamma=.01, kernel_x="rbf", gamma_x = .0000001,
                   kernel_xstar="rbf", gamma_xstar = .00001, tol = 1e-10)
    svmp.fit(X_train, XStar, np.array(y_train).reshape(len(X_train),1))
    y_predict = svmp.predict(X_test)
    correct = np.sum(y_predict == y_test)
    print("Prediction accuracy of SVM")
    print("%d out of %d predictions correct" % (correct, len(y_predict)))




def testSVMPlus():
    #X_train, X_test, y_train, y_test, XStar = 2*np.eye(3), 3*np.eye(3), [1,1,-1], [1,1,-1], 2*np.eye(3)
    #X1 = np.array([[3, 1], [3, -1], [6, 1], [6, -1]])
    #X2 = np.array([[1, 0], [0, 1], [0, -1], [-1, 0]])
    #X_train = XStar = np.concatenate((X1, X2))
    #y_train = [1, 1, 1, 1, -1, -1, -1, -1]


    X_train = np.array([[17, 24,  1,  8, 15],
                       [23,  5,  7, 14, 16],
                       [4,  6, 13, 20, 22],
                       [10, 12, 19, 21,  3],
                       [11, 18, 25,  2,  9]])

    y_train = [1, 1, -1, 1, -1]

    XStar = np.eye(5)
    XStar[0,0] = 1
    XStar[1, 1] = 2
    XStar[2, 2] = 3
    XStar[3, 3] = 4
    XStar[4, 4] = 5
    # train and predict using SVM plus
    #svmp = SVMPlus(C=1, gamma=1, kernel_x= "rbf", gamma_x = .0019,
    #               kernel_xstar="rbf", gamma_xstar = 0.0568)
    #svmp.fit(X_train, XStar, y_train)

    svmp = svmplus.SVMPlus(svm_type="QP", C=1, gamma=1, kernel_x= "linear", degree_x = 2,  gamma_x = .0019,
                   kernel_xstar="linear", degree_xstar = 2, gamma_xstar = 0.0568)
    svmp.fit(X_train, XStar, np.array(y_train).reshape(len(X_train),1))
    #y_predict = svmp.predict(X_test)
    #print(y_predict)


def test3Class():
    X1 = np.array([[3,1], [3,-1], [5,1], [5,-1]])
    X2 = np.array([[1,0], [0,1], [0,-1], [-1,0]])
    X3 = np.array([[1,10], [3,11]])

    X_train = np.concatenate((X1,X2))
    X_train = np.concatenate((X_train, X3))

    y_train = np.array([1,1,1,1,2,2,2,2,3,3]).reshape(10, 1)
    XStar = np.eye(10)
    X_test = np.array([[4,1], [5,0], [4,0], [4,1],
                 [0, 0], [-1,-2], [1,2], [1,-2]])
    y_test = np.array([1,1,1,1, 2, 2, 2, 2]).reshape(8, 1)

    svmp = svmplus.SVMPlus(svm_type="QP", C=1000, gamma=.1, kernel_x="rbf", degree_x=2, gamma_x=.00001,
                   kernel_xstar="rbf", degree_xstar=2, gamma_xstar=0.00001)

    #svmp = libsvm.LibSVMPlus(C=100, gamma=.0001, kernel_x="rbf", degree_x=2, gamma_x=.0001,
    #               kernel_xstar="rbf", degree_xstar=2, gamma_xstar=0.00001)
    svmp.fit(X_train, XStar, np.array(y_train).reshape(len(X_train),1))
    y_predict = svmp.predict(X_test)
    print(y_predict)



if __name__ == "__main__":
    #testSVMPlus()
    #testLinearSVMPlus()
    #testPolynomialSVMPlus()
    #testRbfSVMPlus()
    #testGridSerachCV()
    test3Class()


svmplus/base.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import six
from abc import ABCMeta
import numpy as np
from sklearn.base import BaseEstimator
from numpy import linalg as LA


class BaseSVMPlus(six.with_metaclass(ABCMeta, BaseEstimator)):
    """Base class for SVM plus classification
    """

    def __init__(self, C, gamma,
                 kernel_x, degree_x, gamma_x,
                 kernel_xstar, degree_xstar, gamma_xstar,
                 tol):


        if gamma == 0:
            msg = ("The gamma value of 0.0 is invalid. Use 'auto' to set"
                   " gamma to a value of 1 / n_features.")
            raise ValueError(msg)

        self.C = C
        self.gamma = gamma
        self.kernel_x = kernel_x
        self.degree_x = degree_x
        self.gamma_x = gamma_x
        self.kernel_xstar = kernel_xstar
        self.degree_xstar = degree_xstar
        self.gamma_xstar = gamma_xstar
        self.tol = tol

    def fit(self, X, XStar, y):
        print("fit function")


    def project(self, X):
        print("project function")


    def predict(self, X):
        print("predict function")


    def decision_function(self, X):
        print("decision function")



    @staticmethod
    # Linear kernel
    def _linear_kernel(x1, x2, param = None):
        return np.dot(x1, x2)


    @staticmethod
    # Polynomial kernel
    def _poly_kernel(x1, x2, param = 2):
        return (1 + np.dot(x1, x2)) ** param


    @staticmethod
    # Radial basis kernel
    def _rbf_kernel(x1, x2, param = 1.0):
        return np.exp(-(LA.norm(x1 - x2) ** 2) * param)


svmplus/svmplus.py
# -*- coding: utf-8 -*-

"""Main module."""

from svmplus.svmplusQP import QPSVMPlus
from svmplus.libsvmplus import LibSVMPlus

def SVMPlus(svm_type = "QP", C=1, gamma=1,
                 kernel_x = 'rbf', degree_x = 3, gamma_x ='auto',
                 kernel_xstar = 'rbf', degree_xstar = 3, gamma_xstar ='auto',
                 tol = 1e-10):
    """
        SVMPlus for classification problems.

        Creates and returns an instance of the class specified
        in the svm_type.

        We refer to [1] for theoretical details of LUPI and SVM+,
        and we refer to [2] for implementation details of SVM+ in MATLAB.

        Parameters
        ----------
        svm_type : str
            optimization techiniques based on : QP, LibSVM, LibLinear etc.
            Currently it supports only "QP" and "LIBSVM".
        C : int
            cost of constraints violation
        gamma : int
            parameter needed for priviledged information
        kernel_x : str
            the kernel used for standard training data
        degree_x :
            parameter needed for polynomial kernel for training data
        gamma_x :
            parameter needed for rbf kernel for training data
        kernel_xstar :
            the kernel used for priviledged information (PI)
        degree_xstar :
            parameter needed for polynomial kernel for PI
        gamma_xstar :
            parameter needed for rbf kernel for PI
        tol :
            tolerance of dual variables

        Returns
        -------
        svmp
            an instance of the class specified in the svm_type.

        References
        ----------
        .. [1] Vladimir et. al, Neural Networks, 2009, 22, pp 544–557.
                \url{https://doi.org/10.1016/j.neunet.2009.06.042}

        .. [2] Li et. al, 2016.
            \url{https://github.com/okbalefthanded/svmplus_matlab}

        Examples
        --------
        >>> import numpy as np
        >>> import svmplus
        >>> X1 = np.array([[3,1], [3,-1], [5,1], [5,-1]])
        >>> X2 = np.array([[1,0], [0,1], [0,-1], [-1,0]])
        >>> X3 = np.array([[1,10], [3,11]])
        >>> X_train = np.concatenate((X1,X2))
        >>> X_train = np.concatenate((X_train, X3))
        >>> y_train = np.array([1,1,1,1,2,2,2,2,3,3]).reshape(10, 1)
        >>> XStar = np.eye(10)
        >>> X_test = np.array([[4,1], [5,0], [4,0], [4,1],
        >>>              [0, 0], [-1,-2], [1,2], [1,-2]])
        >>> y_test = np.array([1,1,1,1, 2, 2, 2, 2]).reshape(8, 1)

        >>> svmp = svmplus.SVMPlus(svm_type="QP", C=1000, gamma=.1, kernel_x="rbf", degree_x=2, gamma_x=.00001,
        >>>                kernel_xstar="rbf", degree_xstar=2, gamma_xstar=0.00001)

        >>> svmp.fit(X_train, XStar, np.array(y_train).reshape(len(X_train),1))
        >>> y_predict = svmp.predict(X_test)

        """
    if svm_type == "QP":
        svmp = QPSVMPlus(C, gamma,
                         kernel_x, degree_x, gamma_x,
                         kernel_xstar, degree_xstar, gamma_xstar,
                         tol)
    elif svm_type == "LIBSVM":
        svmp = LibSVMPlus(C, gamma,
                         kernel_x, degree_x, gamma_x,
                         kernel_xstar, degree_xstar, gamma_xstar,
                         tol)
    else:
        print("SVM type not supported")
    return svmp


svmplus/svmplusQP.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Multiclass SVM Plus Classification(SVPC)
https://github.com/TBD
niharika gauraha<niharika.gauraha@farmbio.uu.se>
Ola Spjuth<ola.spjuth@farmbio.uu.se>
Learns a binary classifier based on SVM Plus:
LUPI paradigm. Uses scikit learn.
Licensed under a Creative Commons Attribution-NonCommercial 4.0
International License.
Based on SVM+ by Vapnik et al.
"""


import six
from abc import ABCMeta
import numpy as np
from sklearn.base import BaseEstimator
from cvxopt import matrix, solvers
from base  import BaseSVMPlus
from sklearn.utils import check_X_y


class QPSVMPlus(six.with_metaclass(ABCMeta, BaseSVMPlus, BaseEstimator)):
    def __init__(self, C=1, gamma=1,
                 kernel_x = 'rbf', degree_x = 3, gamma_x ='auto',
                 kernel_xstar = 'rbf', degree_xstar = 3, gamma_xstar ='auto',
                 tol = 1e-5):

        super(QPSVMPlus, self).__init__(C, gamma,
                 kernel_x, degree_x, gamma_x,
                 kernel_xstar, degree_xstar, gamma_xstar,
                 tol)


    def fit(self, X, XStar, y):
        """Fit the SVM model according to the given training data.
                """
        n_samples, n_features = X.shape
        X, y = check_X_y(X, y.flat, 'csr')
        XStar, y = check_X_y(XStar, y.flat, 'csr')
        y = y.reshape(n_samples, 1)


        if self.kernel_x == "linear":
            kernel_method = self._linear_kernel
            kernel_param = None
        elif self.kernel_x == "poly":
            kernel_method = self._poly_kernel
            kernel_param = self.degree_x
        else:
            kernel_method = self._rbf_kernel
            if self.gamma_x == 'auto':
                self.gamma_x = 1 / n_features
            kernel_param = self.gamma_x

        if self.kernel_xstar == "linear":
            kernel_method_star = self._linear_kernel
            kernel_param_star = None
        elif self.kernel_xstar == "poly":
            kernel_method_star = self._poly_kernel
            kernel_param_star = self.degree_xstar
        else:
            kernel_method_star = self._rbf_kernel
            if self.gamma_xstar == 'auto':
                self.gamma_xstar = 1 / XStar.shape[1]
            kernel_param_star = self.gamma_xstar

        # compute the matrix K and KStar (n_samples X n_samples) using kernel function
        K = np.zeros((n_samples, n_samples))
        KStar = np.zeros((n_samples, n_samples))

        for i in range(n_samples):
            for j in range(n_samples):
                K[i, j] = kernel_method(X[i, :], X[j, :], kernel_param)
                KStar[i, j] = kernel_method_star(XStar[i, :], XStar[j, :], kernel_param_star)

        cls_labels = np.unique(y)
        n_class = len(cls_labels)

        if n_class == 2:
            n_class = 1

        self.n_class = n_class
        self.model = []

        for i in range(n_class):
            y_temp = np.array(y)
            y[y != (i + 1)] = -1
            y[y == (i + 1)] = 1

            P1 = np.concatenate((np.multiply(K, y * np.transpose(y)) + KStar / float(self.gamma),
                                 KStar / float(self.gamma)), axis=1)

            P2 = np.concatenate((KStar / float(self.gamma), KStar / float(self.gamma)), axis=1)
            P = np.concatenate((P1, P2), axis=0)
            A = np.concatenate((np.ones((2 * n_samples, 1)),
                                np.concatenate((y, np.zeros((n_samples, 1))))), axis=1)
            A = A.T
            b = np.array([[n_samples * self.C], [0]])
            G = -np.eye(2 * n_samples)
            h = np.zeros((2 * n_samples, 1))

            Q = - (self.C / (self.gamma)) * np.sum(KStar, axis=1)  # [sum(KStar[:,i]) for i in range(n_samples)]
            q = np.concatenate((-1 + Q, Q))

            sol = solvers.qp(matrix(P, tc='d'), matrix(q, tc='d'), matrix(G, tc='d'), matrix(h, tc='d'),
                             matrix(A, tc='d'), matrix(b, tc='d'))
            # Lagrange multipliers
            alpha = np.array(sol['x'][0:n_samples])
            beta = sol['x'][n_samples:2 * n_samples]

            # compute b_star first
            wxstar = (1 / self.gamma) * np.matmul(KStar, (alpha + beta - self.C))
            wxstar_idx = np.ravel(beta) > self.tol

            if (~wxstar_idx).all():  # no beta > tol
                b_star = max(-wxstar)
            else:
                b_star = np.mean(-wxstar[wxstar_idx])

            # Support vectors have non zero lagrange multipliers
            sv = np.ravel(alpha) > self.tol  # tolerance
            sv_x = X[sv]
            sv_y = y[sv]

            d = np.multiply(y, alpha)
            wx = np.matmul(K, d)
            temp = np.multiply(y, (1 - wxstar - b_star))
            wx = -wx + temp
            if (~sv).all():  # no alpha > tol
                lb = np.max(wx[y == 1])
                ub = np.min(wx[y == -1])
                b = (lb + ub) / 2
            else:
                b = np.mean(wx[sv])

            dual_coef = np.multiply(sv_y, alpha[sv])
            intercept = b
            support_vectors = sv_x  # support vector's features
            m = [support_vectors, dual_coef, intercept]

            if n_class == 1:
                self.model = m
            else:
                if i == 0:
                    self.model = list([m])
                else:
                    self.model.append(m)  # list(self.model, [m])

            y = y_temp[:]



    def project(self, X):
        if self.kernel_x == "linear":
            kernel_method = self._linear_kernel
            kernel_param = None
        elif self.kernel_x == "poly":
            kernel_method = self._poly_kernel
            kernel_param = self.degree_x
        else:
            kernel_method = self._rbf_kernel
            kernel_param = self.gamma_x


        if self.n_class == 1:
            y_project = np.zeros(len(X))
            m = self.model
            for i in range(len(X)):
                s = 0
                for a, sv in zip(m[1], m[0]):
                    s += np.multiply(a, kernel_method(X[i], sv, kernel_param))
                y_project[i] = s
        else:
            y_project = np.zeros((len(X), self.n_class))
            for cls in range(self.n_class) :
                m = self.model[cls]
                for i in range(len(X)):
                    s = 0
                    for a, sv in zip(m[1], m[0]):
                        s += np.multiply(a, kernel_method(X[i], sv, kernel_param))
                    y_project[i, cls] = s
        return y_project



    def predict(self, X):
        if self.n_class == 1:
            y_predict = np.sign(self.project(X) +self.model[2])
        else:
            y_predict = -1 * np.ones((len(X), 1))
            y_project = self.project(X)
            for cls in range(self.n_class):
                y_temp = y_project[:, cls] + self.model[cls][2]
                y_predict[y_temp >= 0] = cls+1

        return y_predict


    def decision_function(self, X):
        return self.project(self, X)




svmplus/libsvmplus.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import six
from abc import ABCMeta
import numpy as np
from sklearn.base import BaseEstimator, ClassifierMixin
from sklearn import svm
from sklearn.utils import check_X_y
from base  import BaseSVMPlus


class LibSVMPlus(six.with_metaclass(ABCMeta, BaseSVMPlus, BaseEstimator)):
    """Base class for SVM plus classification
    """

    #_kernels = ["linear", "poly", "rbf"]

    def __init__(self, C=1, gamma=1,
                 kernel_x = 'rbf', degree_x = 3, gamma_x ='auto',
                 kernel_xstar = 'rbf', degree_xstar = 3, gamma_xstar ='auto',
                 tol = 1e-5):

        super(LibSVMPlus, self).__init__(C, gamma,
                                      kernel_x, degree_x, gamma_x,
                                      kernel_xstar, degree_xstar, gamma_xstar,
                                      tol)


    def fit(self, X, XStar, y):
        """Fit the SVM model according to the given training data.
        """
        y = np.array(y).reshape(len(X), 1)
        X, y = check_X_y(X, y.flat, 'csr')
        XStar, y = check_X_y(XStar, y.flat, 'csr')
        n_samples, n_features = X.shape
        y = y.reshape(n_samples, 1)

        if self.kernel_x == "linear":
            kernel_method = self._linear_kernel
            kernel_param = None
        elif self.kernel_x == "poly":
            kernel_method = self._poly_kernel
            kernel_param = self.degree_x
        else:
            kernel_method = self._rbf_kernel
            if self.gamma_x == 'auto':
                self.gamma_x = 1 / n_features
            kernel_param = self.gamma_x

        if self.kernel_xstar == "linear":
            kernel_method_star = self._linear_kernel
            kernel_param_star = None
        elif self.kernel_xstar == "poly":
            kernel_method_star = self._poly_kernel
            kernel_param_star = self.degree_xstar
        else:
            kernel_method_star = self._rbf_kernel
            if self.gamma_xstar == 'auto':
                self.gamma_xstar = 1 / XStar.shape[1]
            kernel_param_star = self.gamma_xstar

        # compute the matrix K and KStar (n_samples X n_samples) using kernel function
        K = np.zeros((n_samples, n_samples))
        KStar = np.zeros((n_samples, n_samples))
        for i in range(n_samples):
            for j in range(n_samples):
                K[i, j] = kernel_method(X[i, :], X[j, :], kernel_param)
                KStar[i, j] = kernel_method_star(XStar[i, :], XStar[j, :], kernel_param_star)

        cls_labels = np.unique(y)
        n_class = len(cls_labels)

        if n_class == 2:
            n_class = 1

        self.n_class = n_class
        self.model = []

        for i in range(n_class):
            y_temp = np.array(y)
            y[y != (i + 1)] = -1
            y[y == (i + 1)] = 1

            # append bias
            K = K + 1
            KStar = KStar + 1

            G = np.eye(n_samples) - np.linalg.inv(np.eye(n_samples) + (self.C / self.gamma) * KStar)
            G = (1 / self.C) * G
            H = np.multiply(K, y * np.transpose(y))
            Q = H + G
            #D = np.transpose(range(n_samples))
            #prob = np.concatenate((D, Q), axis = 1)
            model = svm.libsvm.fit(Q, np.ones(n_samples), svm_type=2, nu = 1 / n_samples,
                                   kernel = 'precomputed', tol = self.tol)

            sv_x = X[model[0]]
            sv_y = y[model[0]]

            coeff = model[3]
            dc = abs(coeff[0]).reshape(len(coeff[0]),1)
            dual_coef = np.multiply(sv_y, dc)
            support_vectors = sv_x  # support vector's features
            m = [support_vectors, dual_coef]

            if n_class == 1:
                self.model = m
            else:
                if i == 0:
                    self.model = list([m])
                else:
                    self.model.append(m)  # list(self.model, [m])

            y = y_temp[:]



    def project(self, X):
        if self.kernel_x == "linear":
            kernel_method = self._linear_kernel
            kernel_param = None
        elif self.kernel_x == "poly":
            kernel_method = self._poly_kernel
            kernel_param = self.degree_x
        else:
            kernel_method = self._rbf_kernel
            kernel_param = self.gamma_x


        if self.n_class == 1:
            y_project = np.zeros(len(X))
            m = self.model
            for i in range(len(X)):
                s = 0
                for a, sv in zip(m[1], m[0]):
                    s += a * kernel_method(X[i], sv, kernel_param)
                y_project[i] = s
        else:
            y_project = np.zeros((len(X), self.n_class))
            for cls in range(self.n_class) :
                m = self.model[cls]
                for i in range(len(X)):
                    s = 0
                    for a, sv in zip(m[1], m[0]):
                        s += np.multiply(a, kernel_method(X[i], sv, kernel_param))
                    y_project[i, cls] = s
        return y_project


    def predict(self, X):
        if self.n_class == 1:
            y_predict = np.sign(self.project(X))
        else:
            y_predict = -1 * np.ones((len(X), 1))
            y_project = self.project(X)
            for cls in range(self.n_class):
                y_temp = y_project[:, cls]
                y_predict[y_temp >= 0] = cls+1

        return y_predict


    def decision_function(self, X):
        return self.project(X)




examples/testLibSVM.py
import numpy as np
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
import scipy.ndimage
import sklearn.preprocessing as preproc
import svmplus

def prepareDigitData():
    #digits = load_digits(n_class=4)
    digits = load_digits()
    X = digits.data
    y = digits.target
    #y[y == 0] = -1
    y = y+1

    resizedImage = np.zeros((len(X), 16))
    for i in range(len(X)):
        originalImage = X[i].reshape(8, 8)
        originalImage = scipy.ndimage.zoom(originalImage, 4 / (8), order=0)
        resizedImage[i] = originalImage.reshape(16, )

    XStar = preproc.scale(X, axis = 1)
    X = preproc.scale(resizedImage, axis = 1)

    X_train, X_test, y_train, y_test, indices_train, indices_valid = \
        train_test_split(X, y, range(len(X)), test_size=0.5, stratify = y)

    XStar = XStar[indices_train]
    return X_train, X_test, y_train, y_test, XStar


def testLinearSVMPlus():
    X_train, X_test, y_train, y_test, XStar = prepareDigitData()

    y_test = np.array(y_test)

    svmp = svmplus.SVMPlus(svm_type="LIBSVM", C=10, kernel_x="linear", kernel_xstar="linear", tol=1e-10)
    svmp.fit(X_train, XStar, y_train)
    y_predict = svmp.predict(X_test)
    correct = np.sum(y_predict.flat == y_test)
    print("Prediction accuracy of SVM")
    print("%d out of %d predictions correct" % (correct, len(y_predict)))


def testPolynomial():
    X_train, X_test, y_train, y_test, XStar = prepareDigitData()

    svmp = svmplus.SVMPlus(svm_type="LIBSVM", C=10, gamma=.01, kernel_x="poly",
                   kernel_xstar="poly", tol=1e-10)
    svmp.fit(X_train, XStar, y_train)
    y_predict = svmp.predict(X_test)
    correct = np.sum(y_predict.flat == y_test)
    print("Prediction accuracy of SVM")
    print("%d out of %d predictions correct" % (correct, len(y_predict)))



def testRbfSVMPlus():
    X_train, X_test, y_train, y_test, XStar = prepareDigitData()

    # train and predict using SVM plus
    svmp = svmplus.SVMPlus(svm_type="LIBSVM", C=10, gamma=.00001, kernel_x="rbf", gamma_x=.000001,
                   kernel_xstar="rbf", gamma_xstar=.000001, tol=1e-10)
    svmp.fit(X_train, XStar, y_train)
    y_predict = svmp.predict(X_test)
    correct = np.sum(y_predict.flat == y_test)
    print("Prediction accuracy of SVM")
    print("%d out of %d predictions correct" % (correct, len(y_predict)))



def testSVMPlus():
    # X_train, X_test, y_train, y_test, XStar = 2*np.eye(3), 3*np.eye(3), [1,1,-1], [1,1,-1], 2*np.eye(3)
    # X1 = np.array([[3, 1], [3, -1], [6, 1], [6, -1]])
    # X2 = np.array([[1, 0], [0, 1], [0, -1], [-1, 0]])
    # X_train = XStar = np.concatenate((X1, X2))
    # y_train = [1, 1, 1, 1, -1, -1, -1, -1]

    X_train = np.array([[17, 24, 1, 8, 15],
                        [23, 5, 7, 14, 16],
                        [4, 6, 13, 20, 22],
                        [10, 12, 19, 21, 3],
                        [11, 18, 25, 2, 9]])

    y_train = [1, 1, -1, 1, -1]

    XStar = np.eye(5)
    XStar[0, 0] = 1
    XStar[1, 1] = 2
    XStar[2, 2] = 3
    XStar[3, 3] = 4
    XStar[4, 4] = 5
    # train and predict using SVM plus
    # svmp = SVMPlus(C=1, gamma=1, kernel_x= "rbf", gamma_x = .0019,
    #               kernel_xstar="rbf", gamma_xstar = 0.0568)
    # svmp.fit(X_train, XStar, y_train)

    svmp = svmplus.SVMPlus(svm_type="LIBSVM", C=1, gamma=1, kernel_x="linear", degree_x=2, gamma_x=.0019,
                   kernel_xstar="linear", degree_xstar=2, gamma_xstar=0.0568)
    svmp.fit(X_train, XStar, y_train)
    # y_predict = svmp.predict(X_test)
    # print(y_predict)


def test3Class():
    X1 = np.array([[3, 1], [3, -1], [5, 1], [5, -1]])
    X2 = np.array([[1, 0], [0, 1], [0, -1], [-1, 0]])
    X3 = np.array([[1, 10], [3, 11]])

    X_train = np.concatenate((X1, X2))
    X_train = np.concatenate((X_train, X3))

    y_train = np.array([1, 1, 1, 1, 2, 2, 2, 2, 3, 3]).reshape(10, 1)
    XStar = np.eye(10)
    X_test = np.array([[4, 1], [5, 0], [4, 0], [4, 1],
                       [0, 0], [-1, -2], [1, 2], [1, -2]])
    y_test = np.array([1, 1, 1, 1, 2, 2, 2, 2]).reshape(8, 1)

    svmp = svmplus.SVMPlus(svm_type = "LIBSVM", C=10, gamma=.01, kernel_x="rbf", degree_x=3, gamma_x=.001,
                   kernel_xstar="rbf", degree_xstar=3, gamma_xstar=0.001, tol = 1e-10)
    svmp.fit(X_train, XStar, y_train)
    y_predict = svmp.predict(X_test)
    print(y_predict)
    correct = np.sum(y_predict == y_test)
    print("Prediction accuracy of SVM")
    print("%d out of %d predictions correct" % (correct, len(y_predict)))


if __name__ == "__main__":
    #testSVMPlus()
    #testLinearSVMPlus()
    #testPolynomial()
    #testRbfSVMPlus()
    test3Class()